File: infrastructure/db-core/src/main/java/com/nmh/commerce/coupon/IssuedCouponEntity.java
package com.nmh.commerce.coupon;

import com.nmh.commerce.BaseEntity;
import com.nmh.commerce.user.UserEntity;
import jakarta.persistence.Entity;
import jakarta.persistence.ManyToOne;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;

@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@Entity
public class IssuedCouponEntity extends BaseEntity {
    @ManyToOne
    private CouponEntity coupon;

    @ManyToOne
    private UserEntity user;

    private LocalDateTime issuedAt;

    @Builder
    private IssuedCouponEntity(CouponEntity coupon, UserEntity user, LocalDateTime issuedAt) {
        this.coupon = coupon;
        this.user = user;
        this.issuedAt = issuedAt;
    }

    public static IssuedCouponEntity from(IssuedCoupon issuedCoupon) {
        return IssuedCouponEntity.builder()
            .coupon(CouponEntity.from(issuedCoupon.coupon))
            .user(UserEntity.from(issuedCoupon.user))
            .issuedAt(issuedCoupon.issuedAt)
            .build();
    }

    public IssuedCoupon toDomain() {
        return new IssuedCoupon(getId(), coupon.toDomain(), user.toDomain(), issuedAt);
    }
}

File: infrastructure/db-core/src/main/kotlin/com/nmh/commerce/coupon/IssuedCouponEntity.kt
package com.nmh.commerce.coupon

import com.nmh.commerce.BaseEntity
import com.nmh.commerce.user.UserEntity
import jakarta.persistence.Entity
import jakarta.persistence.ManyToOne
import java.time.LocalDateTime

@Entity
class IssuedCouponEntity private constructor(
    override val id: Long = 0,
    @field:ManyToOne val coupon: CouponEntity,
    @field:ManyToOne val user: UserEntity,
    val issuedAt: LocalDateTime,
) : BaseEntity<Long>() {
    fun toDomain(): IssuedCoupon = IssuedCoupon(id, coupon.toDomain(), user.toDomain(), issuedAt)

    companion object {
        fun from(issuedCoupon: IssuedCoupon): IssuedCouponEntity =
            IssuedCouponEntity(
                issuedCoupon.id,
                CouponEntity.from(issuedCoupon.coupon),
                UserEntity.from(issuedCoupon.user),
                issuedCoupon.issuedAt,
            )
    }
}
